[inductor] Lower divide by constant as multiplication by reciprocal

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122736
* #122915
* #115435
* #122518
* __->__ #121924

Fixes #101039

This lowers division by a constant value to be multipication by reciprocal.
The same optimization is applied in eager mode on CUDA:

https://github.com/pytorch/pytorch/blob/0636c11811e15d1919cdd6cf20cb2d2bed2ee1da/aten/src/ATen/native/cuda/BinaryDivTrueKernel.cu#L36-L38

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Rebase

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Lower divide by constant as multiplication by reciprocal"


Fixes #101039

This lowers division by a constant value to be multipication by reciprocal.
The same optimization is applied in eager mode on CUDA:

https://github.com/pytorch/pytorch/blob/0636c11811e15d1919cdd6cf20cb2d2bed2ee1da/aten/src/ATen/native/cuda/BinaryDivTrueKernel.cu#L36-L38

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Lower divide by constant as multiplication by reciprocal"


Fixes #101039

This lowers division by a constant value to be multipication by reciprocal.
The same optimization is applied in eager mode on CUDA:

https://github.com/pytorch/pytorch/blob/0636c11811e15d1919cdd6cf20cb2d2bed2ee1da/aten/src/ATen/native/cuda/BinaryDivTrueKernel.cu#L36-L38

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Lower divide by constant as multiplication by reciprocal"


Fixes #101039

This lowers division by a constant value to be multipication by reciprocal.
The same optimization is applied in eager mode on CUDA:

https://github.com/pytorch/pytorch/blob/0636c11811e15d1919cdd6cf20cb2d2bed2ee1da/aten/src/ATen/native/cuda/BinaryDivTrueKernel.cu#L36-L38

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -5339,6 +5342,35 @@ def mul(a, b):
+        inner_fn_args = x.inner_fn_args()
+    else:
+        size = x.get_size()
+        idx = [sympy_index_symbol(f"i{n}") for n, s in enumerate(size)]
```
```suggestion
        idx = [sympy_index_symbol(f"i{n}") for n in range(len(size))]
```

lezcano replied:
Even more, I see this pattern in quite a few places. Do we want to expose it in a coherent API all throughout the IR?

peterbell10 replied:
Yeah it would be useful if all IR nodes had `inner_fn_args`.

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -5339,6 +5342,35 @@ def mul(a, b):
 
+def get_constant_value(x: ir.IRNode) -> Optional[ir.Constant]:
+    """Try convert an arbitrary IR node into an ir.Constant value"""
+    if isinstance(x, ir.MutableBox):
+        return get_constant_value(x.data)
+    if isinstance(x, ir.BaseView):
+        return get_constant_value(x.unwrap_view())
```
Is this just an optimisation? Could we let the code below handle these two cases?

peterbell10 replied:
Yeah, I could just drop these and put everything through `ExtractConstantsHandler` but this avoids creating all the intermediate sympy indexing expressions if they aren't required.

lezcano replied:
Mind leaving a comment?

lezcano commented on torch/_inductor/ops_handler.py:
```diff
@@ -664,3 +664,18 @@ def count(val):
     return h
+
+
+class ExtractConstantsHandler(KernelFormatterHandler):
```
Why do you inherit from `KernelFormatterHandler`?

peterbell10 replied:
This is a workaround for `MockHandler`'s exponential blow-up issue. Basically if you do `x = ops.add(x, x)` repeatedly it expands to a single expression with 2^n elements.

Would be better to add a new ops handler base class that doesn't do any string formatting, but this would be a different change.

lezcano replied:
Do you mind doing that refactoring? This is too hacky otherwise.

lezcano reviewed (commented):
Any reason for not turning this into an actual generic constant folding pass, where anything that's constant is represented by `ir.Constant`?

peterbell10 committed:
Update

[ghstack-poisoned]

lezcano commented:
As a separate point, do we also use this lowering for CPU? A priori, I'd assume we don't

peterbell10 commented:
> Any reason for not turning this into an actual generic constant folding pass, where anything that's constant is represented by `ir.Constant`?

This could certainly be done as an extension, but would be a more complicated change. 

> As a separate point, do we also use this lowering for CPU? A priori, I'd assume we don't

We don't in eager, but this applies the optimization to both CPU and CUDA.

lezcano commented:
>We don't in eager, but this applies the optimization to both CPU and CUDA

I know, but do we want to? If so, we should probably also change CPU eager. cc @jgong5

jgong5 commented:
> I know, but do we want to? If so, we should probably also change CPU eager. cc @jgong5

Yes, we'd better also make the behavior of CPU and CUDA consistent in eager. cc @mingfeima 

@lezcano Perhaps we follow this up after this PR lands?

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Lower divide by constant as multiplication by reciprocal"


Fixes #101039

This lowers division by a constant value to be multipication by reciprocal.
The same optimization is applied in eager mode on CUDA:

https://github.com/pytorch/pytorch/blob/0636c11811e15d1919cdd6cf20cb2d2bed2ee1da/aten/src/ATen/native/cuda/BinaryDivTrueKernel.cu#L36-L38

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented on torch/_inductor/ops_handler.py:
```diff
@@ -494,6 +494,38 @@ def libdevice_log(self, x0: T) -> T:
+
+    @staticmethod
+    def scan(dtypes, combine_fn, values, inits) -> Tuple[None, ...]:
+        return tuple(None for i in range(len(values)))
```
nit.
```suggestion
        return (None,) * len(values)
```
but sure.

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge
